symsrv-fetch.py

```diff
@@ -24,345 +24,468 @@
+    tasks = []
+    for filename, debug_id, code_file, code_id, has_code in modules:
+        tasks.append(
+            dump_module(
+                output,
+                symcache,
+                filename,
+                debug_id,
+                code_file,
+                code_id,
+                has_code,
+                dump_syms,
+            )
+        )
+
+    return await asyncio.gather(*tasks)
+
+
+def dump(output, symcache, modules, dump_syms):
+    res = asyncio.run(dump_helper(output, symcache, modules, dump_syms))
+    file_index = {x for x in res if isinstance(x, str)}
+    stats = {
+        "dump_error": sum(1 for x in res if x == 1),
+        "no_bin": sum(1 for x in res if x == 2),
+    }
+
+    return file_index, stats
+
+
+async def collect_helper(modules):
+    loop = asyncio.get_event_loop()
+    tasks = []
+    connector = TCPConnector(limit=100, limit_per_host=20)
+    async with ClientSession(
+        loop=loop, timeout=ClientTimeout(total=TIMEOUT), connector=connector
+    ) as client:
+        for filename, ids in modules.items():
+            for debug_id, code_file, code_id in ids:
+                tasks.append(
+                    collect_info(client, filename, debug_id, code_file, code_id)
+                )
+
+        return await asyncio.gather(*tasks)
+
+
+def collect(modules):
+    res = asyncio.run(collect_helper(modules))
+    to_dump = []
+    stats = {"no_pdb": 0, "is_there": 0}
+    for filename, debug_id, code_file, code_id, has_pdb, has_code, is_there in res:
+        if not has_pdb:
+            if is_there:
+                stats["is_there"] += 1
+            else:
+                stats["no_pdb"] += 1
+                log.info(f"No pdb for {filename}/{debug_id}")
+            continue
 
-    if not file_index:
         log.info(
-            'No symbols downloaded: %d considered, '
-            '%d already present, %d in blacklist, %d skipped, %d not found, '
-            '%d not processed ' %
-            (total,
-             existing_count,
-             blacklist_count,
-             skiplist_count,
-             not_found_count,
-             not_processed_count))
-        write_skiplist(skiplist)
-        sys.exit(0)
+            f"To dump: {filename}/{debug_id}, {code_file}/{code_id} and has_code = {has_code}"
+        )
+        to_dump.append((filename, debug_id, code_file, code_id, has_code))
+
+    log.info(f"Collected {len(to_dump)} files to dump")
+
+    return to_dump, stats
+
+
+async def fetch_and_write(output, client, filename, file_id):
+    path = os.path.join(filename, file_id, filename)
+    data = await fetch_file(client, MICROSOFT_SYMBOL_SERVER, path)
+
+    if not data:
+        return False
 
-    # Write an index file
-    with zipfile.ZipFile(args.zip, 'w', zipfile.ZIP_DEFLATED) as z:
+    output_dir = os.path.join(output, filename, file_id)
+    if not os.path.exists(output_dir):
+        os.makedirs(output_dir)
```

Comment:
Nit: os.makedirs(output_dir, exist_ok=True) so you don't need to check for existance.

Note that this I/O will not be async, so it will block all tasks. You can do something like https://github.com/mozilla/microannotate/blob/c342c8238126ab297988c7a3478e708918d86612/microannotate/generator.py#L254-L257 to hide the sync I/O in asyncio's thread pool and avoid blocking tasks.